### PR TITLE
Kubelet: parallelize cleaning up containers in unwanted pods

### PR DESCRIPTION
**Pull request by yujuhong**
Kubelet kills unwanted pods in SyncPods, which directly impact the latency of a
sync iteration. This change parallelizes the cleanup to lessen the effect.

Eventually, we should leverage per-pod workers for cleanup, with the exception
of truly orphaned pods.

### CODE REVIEW

**Comment by yujuhong** [comment]
This addresses #6764

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -937,17 +937,25 @@ func (kl *Kubelet) pullImage(img string, ref *api.ObjectReference) error {
...
 	// Send the kills in parallel since they may take a long time.
-	errs := make(chan error, len(runningPod.Containers))
+	errs := make(chan error, len(pod.Containers))
 	wg := sync.WaitGroup{}
-	for _, container := range runningPod.Containers {
+	for _, container := range pod.Containers {
 		wg.Add(1)
 		go func(container *kubecontainer.Container) {
 			defer util.HandleCrash()
+			// call the networking plugin for teardown
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Can we add a TODO to handle this without singling our the pod infra container? Main motivation is for Rocket integration which does not have one. We're working on moving the pod infra to be a Docker-only concept.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -937,17 +937,25 @@ func (kl *Kubelet) pullImage(img string, ref *api.ObjectReference) error {
...
+			if container.Name == dockertools.PodInfraContainerName {
+				err := kl.networkPlugin.TearDownPod(pod.Namespace, pod.Name, dockertools.DockerID(container.ID))
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: This change seems unrelated from the rest of this PR. Is it required by it or can we split it out into its own logical comit/PR?

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1452,48 +1460,70 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
...
-		// Kill all the containers in the unidentified pod.
-		for _, c := range pod.Containers {
-			// call the networking plugin for teardown
-			if c.Name == dockertools.PodInfraContainerName {
-				err := kl.networkPlugin.TearDownPod(pod.Namespace, pod.Name, dockertools.DockerID(c.ID))
-				if err != nil {
-					glog.Errorf("Network plugin pre-delete method returned an error: %v", err)
-				}
+		numWorkers++
+		go func(pod *kubecontainer.Pod, ch chan result) {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Lets break this into a different method, SyncPods() is huge already.

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1452,48 +1460,70 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
...
+	for i := 0; i < numWorkers; i++ {
+		m := <-ch
+		if m.err != nil {
+			errs = append(errs, m.err.Error())
+			continue
+		}
+		running = append(running, m.containers...)
+	}
+	if len(errs) != 0 {
+		err := fmt.Errorf("%s", strings.Join(errs, ", "))
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Consider using errors.Aggregate so we have similar error messages:

https://github.com/yujuhong/kubernetes/blob/para_cleanup/pkg/util/errors/errors.go#L23

**Comment by vmarmol** [comment]
Minor comments, looks good overall. Thanks @yujuhong!

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: I don't think it's unrelated. It was used as part of the logic to kill pods in SyncPods. I moved it here since I was going to resue this function. I am not sure why this function didn't call TearDownPod initially though.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Done

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: I moved the whole block (including result aggregation) to a different function. This should improve the readability a little bit.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Done.

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -1503,6 +1490,67 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
...
+	}
+	return running, utilErrors.NewAggregate(errs)
+}
+
+func (kl *Kubelet) getRunningContainersByPod(pod *kubecontainer.Pod) ([]*docker.Container, error) {
+	containerIDs := make([]string, len(pod.Containers))
+	for i, c := range pod.Containers {
+		containerIDs[i] = string(c.ID)
+	}
+	return kl.containerManager.GetRunningContainers(containerIDs)
```
Comment (yifan-gu) [inline pkg/kubelet/kubelet.go]: Thought not related to this PR, but I always have a question about this, I am not sure why the running containers are necessary for removing volumes. I suspect that if the volume is still being used by some container, TearDown() will just return an error (such as device busy).I think if we can remove this logic, it would be great.

As I can remember, this is part of a fix for volume leaks, but I am not sure the original problem is caused by deleting alive volumes..

Let me find out the issue number...

Comment (yifan-gu) [inline pkg/kubelet/kubelet.go]: #3965 #4076 #4086

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Ah I see what you mean.

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: I believe that in theory we should be able to remove it and the volume plugins would handle it. However, it seems like we've had quite a few bugs in this area which lead to data loss. That is what has kept the current logic I believe.

**Comment by vmarmol** [comment]
LGTM, we can squash and merge

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (yifan-gu) [inline pkg/kubelet/kubelet.go]: ack

**Comment by yujuhong** [comment]
Squashed. Thanks!
